### PR TITLE
fix redundant checkpointing in example training scripts

### LINKED ISSUE #32653: redundant checkpointing in example scripts
### System Info

- `transformers` version: 4.43.4
- Platform: Linux-5.14.0-284.40.1.el9_2.x86_64-x86_64-with-glibc2.34
- Python version: 3.12.4
- Huggingface_hub version: 0.24.5
- Safetensors version: 0.4.4
- Accelerate version: 0.33.0
- Accelerate config: 	not found
- PyTorch version (GPU?): 2.4.0 (False)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: No

### Who can help?

@arth

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

In several of the example scripts, running with ``gradient_accumulation_steps > 1`` causes redundant checkpointing. For example, [here](https://github.com/huggingface/transformers/blob/481e15604a0527fec3cbdcccd350f9374a9116b1/examples/pytorch/language-modeling/run_clm_no_trainer.py#L641):
```python 
if completed_steps % checkpointing_steps == 0:
```
This will keep needlessly checkpointing the accelerator state ``gradient_accumulation_steps`` times until the ``completed_steps`` is iterated again. This can be particularly problematic when ``gradient_accumulation_steps`` is large.

The same issue affects several of the example training scripts: _e.g._ ``run_clm_no_trainer``, ``run_fim_no_trainer``, ``run_mlm_no_trainer``, _etc._ (and probably more).

### Expected behavior

A quick fix is to add a clause to the if statement above to make sure checkpointing is done only once when ``gradient_accumulation_steps > 1``, _e.g._:
```python 
if completed_steps % checkpointing_steps == 0 and step % args.gradient_accumulation_steps == 0:
```
where ``step`` is the dataloader step.

### PR DESCRIPTION
**Pull request by eminorhan**
# What does this PR do?

Fixes #32653.

Briefly, in several of the example training scripts, running with `gradient_accumulation_steps > 1` currently causes `gradient_accumulation_steps` times redundant checkpointing in the step-based checkpointing mode. This PR fixes the issue by adding a clause to the step-based checkpointing condition to make sure saving is done only once at the appropriate checkpointing step.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@ArthurZucker

### CODE REVIEW

**Review by LysandreJik — APPROVED** [review]
Thanks for your PR @eminorhan!

Just a quick ping @SunMarc as you work with the `Trainer` and example scripts much more than I do and might see something I don't

**Review by SunMarc — APPROVED** [review]
Thanks for finding and fixing the issue @eminorhan ! I left a suggestion to propagate to other scripts if you think it is better !

File: examples/pytorch/image-classification/run_image_classification_no_trainer.py [hunk 1]
```diff
@@ -544,7 +544,7 @@ def collate_fn(examples):
                 completed_steps += 1
 
             if isinstance(checkpointing_steps, int):
-                if completed_steps % checkpointing_steps == 0:
+                if completed_steps % checkpointing_steps == 0 and step % args.gradient_accumulation_steps == 0:
```
[the remark this answers is not in the document]
Comment (eminorhan) [reply examples/pytorch/image-classification/run_image_classification_no_trainer.py]: @SunMarc your suggestion looks good to me, but I just noticed after committing your change that it doesn't change all scripts. Is there a quick way to fix this?

File: examples/pytorch/image-pretraining/run_mim_no_trainer.py [hunk 2]
```diff
@@ -723,7 +723,7 @@ def preprocess_images(examples):
                 completed_steps += 1
 
             if isinstance(checkpointing_steps, int):
-                if completed_steps % checkpointing_steps == 0:
+                if completed_steps % checkpointing_steps == 0 and step % args.gradient_accumulation_steps == 0:
```
Comment (SunMarc) [inline examples/pytorch/image-pretraining/run_mim_no_trainer.py:726]:
```suggestion
                if completed_steps % checkpointing_steps == 0 and accelerator.sync_gradients:
```

File: examples/pytorch/instance-segmentation/run_instance_segmentation_no_trainer.py [hunk 3]
```diff
@@ -639,7 +639,7 @@ def main():
                 completed_steps += 1
 
             if isinstance(checkpointing_steps, int):
-                if completed_steps % checkpointing_steps == 0:
+                if completed_steps % checkpointing_steps == 0 and step % args.gradient_accumulation_steps == 0:
```
Comment (SunMarc) [inline examples/pytorch/instance-segmentation/run_instance_segmentation_no_trainer.py:642]:
```suggestion
                if completed_steps % checkpointing_steps == 0 and accelerator.sync_gradients:
```

File: examples/pytorch/language-modeling/run_clm_no_trainer.py [hunk 4]
```diff
@@ -638,7 +638,7 @@ def group_texts(examples):
                 completed_steps += 1
 
             if isinstance(checkpointing_steps, int):
-                if completed_steps % checkpointing_steps == 0:
+                if completed_steps % checkpointing_steps == 0 and step % args.gradient_accumulation_steps == 0:
```
Comment (SunMarc) [inline examples/pytorch/language-modeling/run_clm_no_trainer.py:641]:
```suggestion
                if completed_steps % checkpointing_steps == 0 and accelerator.sync_gradients:
```

File: examples/pytorch/language-modeling/run_mlm_no_trainer.py [hunk 5]
```diff
@@ -675,7 +675,7 @@ def group_texts(examples):
                 completed_steps += 1
 
             if isinstance(checkpointing_steps, int):
-                if completed_steps % checkpointing_steps == 0:
+                if completed_steps % checkpointing_steps == 0 and step % args.gradient_accumulation_steps == 0:
```
Comment (SunMarc) [inline examples/pytorch/language-modeling/run_mlm_no_trainer.py:678]:
```suggestion
                if completed_steps % checkpointing_steps == 0 and accelerator.sync_gradients:
```

File: examples/pytorch/object-detection/run_object_detection_no_trainer.py [hunk 6]
```diff
@@ -677,7 +677,7 @@ def main():
                 completed_steps += 1
 
             if isinstance(checkpointing_steps, int):
-                if completed_steps % checkpointing_steps == 0:
+                if completed_steps % checkpointing_steps == 0 and step % args.gradient_accumulation_steps == 0:
```
Comment (SunMarc) [inline examples/pytorch/object-detection/run_object_detection_no_trainer.py:680]:
```suggestion
                if completed_steps % checkpointing_steps == 0 and accelerator.sync_gradients:
```

File: examples/pytorch/question-answering/run_qa_beam_search_no_trainer.py [hunk 7]
```diff
@@ -879,7 +879,7 @@ def create_and_fill_np_array(start_or_end_logits, dataset, max_len):
                 completed_steps += 1
 
             if isinstance(checkpointing_steps, int):
-                if completed_steps % checkpointing_steps == 0:
+                if completed_steps % checkpointing_steps == 0 and step % args.gradient_accumulation_steps == 0:
```
Comment (SunMarc) [inline examples/pytorch/question-answering/run_qa_beam_search_no_trainer.py:882]:
```suggestion
                if completed_steps % checkpointing_steps == 0 and accelerator.sync_gradients:
```

File: examples/pytorch/question-answering/run_qa_no_trainer.py [hunk 8]
```diff
@@ -894,7 +894,7 @@ def create_and_fill_np_array(start_or_end_logits, dataset, max_len):
                 completed_steps += 1
 
             if isinstance(checkpointing_steps, int):
-                if completed_steps % checkpointing_steps == 0:
+                if completed_steps % checkpointing_steps == 0 and step % args.gradient_accumulation_steps == 0:
```
Comment (SunMarc) [inline examples/pytorch/question-answering/run_qa_no_trainer.py:897]:
```suggestion
                if completed_steps % checkpointing_steps == 0 and accelerator.sync_gradients:
```

File: examples/pytorch/semantic-segmentation/run_semantic_segmentation_no_trainer.py [hunk 9]
```diff
@@ -516,7 +516,7 @@ def preprocess_batch(example_batch, transforms: A.Compose):
                 completed_steps += 1
 
             if isinstance(checkpointing_steps, int):
-                if completed_steps % checkpointing_steps == 0:
+                if completed_steps % checkpointing_steps == 0 and step % args.gradient_accumulation_steps == 0:
```
Comment (SunMarc) [inline examples/pytorch/semantic-segmentation/run_semantic_segmentation_no_trainer.py:519]:
```suggestion
                if completed_steps % checkpointing_steps == 0 and accelerator.sync_gradients:
```

File: examples/pytorch/summarization/run_summarization_no_trainer.py [hunk 10]
```diff
@@ -688,7 +688,7 @@ def postprocess_text(preds, labels):
                 completed_steps += 1
 
             if isinstance(checkpointing_steps, int):
-                if completed_steps % checkpointing_steps == 0:
+                if completed_steps % checkpointing_steps == 0 and step % args.gradient_accumulation_steps == 0:
```
Comment (SunMarc) [inline examples/pytorch/summarization/run_summarization_no_trainer.py:691]:
```suggestion
                if completed_steps % checkpointing_steps == 0 and accelerator.sync_gradients:
```

File: examples/pytorch/token-classification/run_ner_no_trainer.py [hunk 11]
```diff
@@ -722,7 +722,7 @@ def compute_metrics():
                 completed_steps += 1
 
             if isinstance(checkpointing_steps, int):
-                if completed_steps % checkpointing_steps == 0:
+                if completed_steps % checkpointing_steps == 0 and step % args.gradient_accumulation_steps == 0:
```
Comment (SunMarc) [inline examples/pytorch/token-classification/run_ner_no_trainer.py:725]:
```suggestion
                if completed_steps % checkpointing_steps == 0 and accelerator.sync_gradients:
```

**Review by SunMarc — APPROVED** [review]
Here you go ! I added the remaining suggestion

**Pull request merged**